### ISSUE TITLE
Ability to duplicate a worksheet?

### ISSUE THREAD

**Issue by jnizniowski** [burnash/gspread#452]
I'm not sure if API makes it possible, but I'm looking for an option to duplicate a worksheet inside the spreadsheet. It would be great to see it in Gspread.

**Comment by Sixninetwo** [comment]
I would also love to see this feature.

**Comment by wlmbgm** [comment]
+1 this would be very helpful.

**Comment by MikhailSpirin** [comment]
+1! But for now will do it with code

**Comment by burnash** [comment]
@robsgithub there's a PR for this. I'll refer it here #459 

In this PR @wtluke suggests this api:

`n = gc.create(copy='yes',key='key/fileID',title='New spreadsheet (copy)')`

While @Drachenfels thinks making it as `gc.copy()` is a better idea.

I'd like to ask public opinion for what would be the best API. Without a flame war, please ;)

**Comment by wtluke** [comment]
Pick one and go for it. This PR was from May last year...!

**Comment by burnash** [comment]
@wtluke I'm sorry that it takes a long time to merge a PR. But you probably know that it's more than just hitting a 'Merge' button. 

I need to test it manually, I need to make sure it has a unit test suite (#459 does not). I need to make sure that all unit tests pass. As a project maintainer, I need to make sure that the new feature in the API is as consistent as possible with the rest of API. 

And most importantly, the project maintainer needs to understand that once the new feature is merged _it's the the project maintainer_ who is responsible for keeping this feature in a working state and answering incoming issues. This is why it's so important to pick the right API.

That's a lot of work. It's all done in my spare time. So yes, it can take long, again sorry about that.

**Comment by robsgithub** [comment]
@burnash  I like the method call idea better as well- gc.copy()
Also thanks a ton for maintaining this repo! I use it constantly for updating my google sheets in a simple way. This is the first i have not been able to accomplish a task with it.

**Comment by robsgithub** [comment]
Perhaps it could work like this. This could potentially save needing to pass the id since it should already be known by the given worksheet.

```
worksheet1 = gc.open('My poor gym results') 
worksheet2 =worksheet1.copy('Copy of My poor gym results') 
```

**Comment by robsgithub** [comment]
Ive been looking closer at PR #459 and it seems that it in regards to copying entire spreadsheets(including all their worksheets). Unless im mistaken thats not what Im after

I thought this thread was for copying individual worksheets within a spreadsheet
I would imagine the functionality could work like this. 

```
worksheet1 = sh.worksheet('January')
worksheet2 = worksheet1.copy('February') 

```

**Comment by jnizniowski** [comment]
Yes, I meant copying a worksheet, not a whole spreadsheet, as in the @robsgithub example.

**Comment by fionazj** [comment]
Seemed like current this function is not implemented yet..Looking forward to this....

**Comment by scythargon** [comment]
+1

**Comment by burnash** [comment]
@jnizniowski You're right I mixed it up a bit. Added sheets duplication in #617. Tests will be added in a follow-up PR and the feature will be released on PyPI in 3.1.0.

API Docs: [`Worksheet.duplicate()`](https://gspread.readthedocs.io/en/latest/api.html#gspread.models.Worksheet.duplicate)

Thanks for the feature suggestion and thanks for everyone for contributing to the discussion.

**Comment by neichler-upl** [comment]
@hugoba88 
Here you go!

`sht = client.open('spreadsheet')
sht.duplicate_sheet(0,insert_sheet_index=1,new_sheet_id=None,new_sheet_name="newsheet")`

**Comment by punnerud** [comment]
This worked for me. Can't find it documented:
```
import gspread
from oauth2client.service_account import ServiceAccountCredentials

gc = gspread.service_account("credentials.json") #The account also need Drive API access. Will give a warning+help.
sh = gc.copy('your sheetID here',title='New spreadsheet (copy)')
sh.share('your_username@example.com', perm_type='user', role='owner') #Have to share the new sheet created with the ServiceAccount
```
